cmd/moco-init/main.go

```diff
@@ -0,0 +1,209 @@
+		return fmt.Errorf("bad POD_NAME: %s", config.podName)
+	}
+
+	indexUint64, err := strconv.ParseUint(fields[len(fields)-1], 10, 32)
+	if err != nil {
+		return fmt.Errorf("bad POD_NAME %s", config.podName)
+	}
+	config.podIndex = uint32(indexUint64)
+
+	baseUint64, err := strconv.ParseUint(serverIDBase, 10, 32)
+	if err != nil {
+		return fmt.Errorf("invalid server base ID: %s: %w", os.Args[1], err)
+	}
+	config.baseID = uint32(baseUint64)
+
+	if err := validateFlags(); err != nil {
+		return err
+	}
+
+	_, err = os.Stat(filepath.Join(config.dataDir, initializedFile))
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+		if err := initMySQL(mysqld); err != nil {
+			return err
+		}
+	}
+
+	return createConf()
+}
+
+func initMySQL(mysqld string) error {
+	dataDir := filepath.Join(config.dataDir, "data")
+	if err := os.RemoveAll(dataDir); err != nil {
+		return fmt.Errorf("failed to remove dir %s: %w", dataDir, err)
+	}
+
+	var args []string
+	args = append(args, "--basedir="+config.baseDir)
+	args = append(args, "--datadir="+dataDir)
+	args = append(args, "--initialize-insecure")
+
+	// Set only if lower-case-table-names flag is set.
+	if isVisitLowerCaseTableNamesFlag() {
+		args = append(args, fmt.Sprintf("--lower_case_table_names=%d", config.lowerCaseTableNames))
+	}
+
+	cmd := exec.Command(mysqld, args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+
+	dotFile := filepath.Join(config.dataDir, "."+initializedFile)
+	if err := os.Remove(dotFile); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	f, err := os.Create(dotFile)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if err := unix.Syncfs(int(f.Fd())); err != nil {
+		return fmt.Errorf("failed to sync fs: %w", err)
+	}
+
+	if err := os.Rename(dotFile, filepath.Join(config.dataDir, initializedFile)); err != nil {
+		return err
+	}
+
+	g, err := os.OpenFile(config.dataDir, os.O_RDONLY, 0755)
+	if err != nil {
+		return err
+	}
+	defer g.Close()
+	return g.Sync()
+}
+
+func createConf() error {
+	tmpl := template.Must(template.New("my.cnf").Parse(mycnfTmpl))
+
+	v := struct {
+		ServerID     uint32
+		AdminAddress string
+	}{
+		ServerID:     config.baseID + config.podIndex,
+		AdminAddress: config.podName,
+	}
+
+	f, err := os.OpenFile(filepath.Join(config.confDir, "my.cnf"), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to create my.cnf file: %w", err)
+	}
+	defer f.Close()
+
+	if err := tmpl.Execute(f, v); err != nil {
+		return err
+	}
+	return f.Sync()
+}
+
+func validateFlags() error {
+	// refs: https://dev.mysql.com/doc/refman/8.0/en/identifier-case-sensitivity.html
+	switch config.lowerCaseTableNames {
+	case 0, 1, 2:
+		// noop
+	default:
+		return errors.New("the value of lower-case-table-names flag must be 0, 1 or 2")
+	}
+
+	return nil
+}
+
+func isVisitLowerCaseTableNamesFlag() bool {
+	found := false
+	flag.Visit(func(f *flag.Flag) {
```

Comment:
Does this code work as expected? I think we should use `rootCmd.Flags().Visit()`.

I checked `flag.Visit()` with `cobra.Command` by the following code.
But I could not check the existence of the argument.

```go
package main

import (
	"flag"
	"fmt"
	"os"

	"github.com/spf13/cobra"
	"github.com/spf13/pflag"
)

var config struct {
	lowerCaseTableNames int
}

func main() {
	if err := rootCmd.Execute(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}

var rootCmd = &cobra.Command{
	RunE: func(cmd *cobra.Command, args []string) error {
		flag.Visit(func(f *flag.Flag) {
			if f.Name == "lower-case-table-names" {
				fmt.Println("flag.Visit") // Not displayed
			}
		})
		cmd.Flags().Visit(func(f *pflag.Flag) {
			if f.Name == "lower-case-table-names" {
				fmt.Println("cmd.Flags().Visit") // Displayed.
			}
		})
		return nil
	},
}

func init() {
	rootCmd.Flags().IntVar(&config.lowerCaseTableNames, "lower-case-table-names", 0, "The value to pass to the '--lower-case-table-names' flag.")
}
```

Reply:
Sorry, thank you for your suggestion.
I fixed it: https://github.com/cybozu-go/moco-agent/pull/71/commits/ee31a6d386db58bb7cf6bd3f844262690d8b5ffd